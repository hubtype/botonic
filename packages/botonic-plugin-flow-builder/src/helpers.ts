import { Plugin } from '@botonic/core'

import {
  HtHandoffNode,
  HtNodeWithContent,
  HtNodeWithContentType,
} from './content-fields/hubtype-fields'
import BotonicPluginFlowBuilder from './index'

const FLOW_BUILDER_PLUGIN_NAME = 'BotonicPluginFlowBuilder'

export function getFlowBuilderPlugin(plugins: {
  [id: string]: Plugin
}): BotonicPluginFlowBuilder {
  const flowBuilderPlugin = Object.values(plugins).find(
    // @ts-ignore
    plugin => plugin.name.includes(FLOW_BUILDER_PLUGIN_NAME)
  ) as BotonicPluginFlowBuilder
  if (!flowBuilderPlugin)
    throw new Error(
      `You must include '@botonic/plugin-flow-builder' in your plugins file.`
    )
  return flowBuilderPlugin
}

export function isHandoffNode(node: HtNodeWithContent): node is HtHandoffNode {
  return node.type === HtNodeWithContentType.HANDOFF
}
