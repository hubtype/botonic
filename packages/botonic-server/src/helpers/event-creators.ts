import {
  BaseEvent,
  BotonicEvent,
  BotonicMessageEvent,
  ConnectionEventStatuses,
  EventTypes,
  MessageEventAck,
  MessageEventFrom,
  User,
} from '@botonic/core'
import { ulid } from 'ulid'

// TODO: To be obtained from new row in DynamoDB table
export const ID_FROM_CHANNEL = '1234'

export function initBaseEvent(user: User, eventType: EventTypes): BaseEvent {
  return {
    userId: user.id,
    eventId: ulid(),
    createdAt: new Date().toISOString(),
    idFromChannel: user.idFromChannel || ID_FROM_CHANNEL,
    channel: user.channel,
    eventType,
  }
}

type ConnectionEventArgs = { user: User; status: ConnectionEventStatuses }
export function createConnectionEvent(args: ConnectionEventArgs): BotonicEvent {
  const { user, status } = args
  return {
    ...initBaseEvent(user, EventTypes.CONNECTION),
    status,
  } as BotonicEvent
}

type BotonicMessageEventArgs = {
  user: User
  properties: Partial<BotonicMessageEvent>
}
export function createMessageEvent(
  args: BotonicMessageEventArgs
): BotonicEvent {
  const { user, properties } = args
  return {
    ...properties,
    ...initBaseEvent(user, EventTypes.MESSAGE),
  } as BotonicEvent
}

type IntegrationEventArgs = { user: User; details: any }
export function createIntegrationEvent(
  eventType: EventTypes,
  args: IntegrationEventArgs
): BotonicEvent {
  const { user, details } = args
  return { ...initBaseEvent(user, eventType), details } as BotonicEvent
}

function initChannelInformation(user: User) {
  const { idFromChannel, channel } = user
  return { idFromChannel, channel }
}

// TODO: Validate how to define these kind of events
// We need channel information in these events to distinguish in which cases it should be executed
type WebchatActionEventArgs = { user: User; action: string; properties: any }
export function createWebchatActionEvent(args: WebchatActionEventArgs): any {
  const { user, properties, action } = args
  return {
    ...initChannelInformation(user),
    ...properties,
    action,
  } as BotonicEvent
}
