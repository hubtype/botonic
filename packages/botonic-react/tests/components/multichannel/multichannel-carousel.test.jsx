import {
  MultichannelButton,
  Element,
  Pic,
  Subtitle,
  Title,
  Carousel,
  Button,
} from '../../../src'
import {
  Multichannel,
  MultichannelCarousel,
} from '../../../src/components/multichannel'
import React from 'react'
import { whatsappRenderer } from '../../helpers/test-utils'

const movies = [
  {
    name: 'Snatch',
    desc: 'Five minutes, Turkish',
    url: 'https://www.imdb.com/title/tt0208092',
    pic:
      'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
  },
  {
    name: 'El laberinto del fauno',
    desc: '2006',
    url: 'https://www.imdb.com/title/tt0457430',
    pic:
      'https://m.media-amazon.com/images/M/MV5BMDBjOWYyMDQtOWRmOC00MDgxLWIxM2UtMTNjYzFiN2RkYTBlXkEyXkFqcGdeQXVyMTAyOTE2ODg0._V1_UY268_CR6,0,182,268_AL_.jpg',
  },
]

export const LEGACY_CONTEXT = {
  indexSeparator: '.',
  messageSeparator: null,
}

export const LEGACY_PROPS = {
  indexMode: 'number',
  showTitle: true,
  showSubtitle: true,
}

describe('Multichannel carousel COMPACT mode', () => {
  test('dynamic carousel with url buttons', () => {
    const sut = (
      <MultichannelCarousel {...LEGACY_PROPS}>
        {movies.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            <MultichannelButton url={e.url}>Visit website</MultichannelButton>
          </Element>
        ))}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut, LEGACY_CONTEXT)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  const options = [
    {
      name: 'Title1',
      desc: 'Subtitle1',
      payload: 'Payload1',
      buttonText: 'Previo a la compra',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
    },
    {
      name: 'Title2',
      desc: 'Subtitle2',
      payload: 'Payload2',
      buttonText: 'Durante la compra',
      pic:
        'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
    },
  ]

  test('dynamic carousel with payload/path buttons', () => {
    const sut = (
      <MultichannelCarousel {...LEGACY_PROPS}>
        {options.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            <MultichannelButton payload={e.payload}>
              {e.buttonText}
            </MultichannelButton>
          </Element>
        ))}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut, LEGACY_CONTEXT)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  const optionsNoTitleNoSubtitle = [
    {
      name: '',
      desc: '',
      payload: 'Payload1',
      buttonText: 'Previo a la compra',
      pic:
        'https://cdn.slidesharecdn.com/profile-photo-sumeet.moghe-48x48.jpg?cb=1527568614',
    },
    {
      name: '',
      desc: '',
      payload: 'Payload2',
      buttonText: 'Durante la compra',
      pic:
        'https://nebula.wsimg.com/obj/NzQ3QUYxQzZBNzE4NjNFRTc1MTU6NmM4YjgzZWVlZTE2MGMzM2RkMTdlZjdjNGUyZmFhNDE6Ojo6OjA=',
    },
  ]

  test('dynamic carousel (just text in button) with payload/path buttons', () => {
    const sut = (
      <MultichannelCarousel {...LEGACY_PROPS}>
        {optionsNoTitleNoSubtitle.map((e, i) => (
          <Element key={e.name}>
            <Pic src={e.pic} />
            <Title>{e.name}</Title>
            <Subtitle>{e.desc}</Subtitle>
            <MultichannelButton payload={e.payload}>
              {e.buttonText}
            </MultichannelButton>
          </Element>
        ))}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut, LEGACY_CONTEXT)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('all elements in same message', () => {
    const sut = (
      <MultichannelCarousel>
        {[
          <Element key={0}>
            <Subtitle>Subtitle will not appear</Subtitle>
            {[
              <MultichannelButton key={1} payload='Payload1'>
                {'Previo'}
              </MultichannelButton>,
            ]}
          </Element>,
          <Element key={1}>
            <Subtitle>Subtitle will not appear</Subtitle>
            {[
              <MultichannelButton key={1} payload='Payload2'>
                {'Durante'}
              </MultichannelButton>,
            ]}
          </Element>,
        ]}
      </MultichannelCarousel>
    )
    const renderer = whatsappRenderer(sut, { messageSeparator: '\n' })
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
