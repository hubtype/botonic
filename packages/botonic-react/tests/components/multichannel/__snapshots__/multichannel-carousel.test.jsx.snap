// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multichannel carousel COMPACT mode dynamic carousel (just text in button) with payload/path buttons 1`] = `
Array [
  <message
    newkey={0}
    type="text"
  >
    
     1. Previo a la compra
  </message>,
  <message
    newkey={1}
    type="text"
  >
    
     2. Durante la compra
  </message>,
]
`;

exports[`Multichannel carousel COMPACT mode dynamic carousel with payload/path buttons 1`] = `
Array [
  <message
    newkey={0}
    type="text"
  >
    *Title1* - _Subtitle1_
    
 1. Previo a la compra
  </message>,
  <message
    newkey={1}
    type="text"
  >
    *Title2* - _Subtitle2_
    
 2. Durante la compra
  </message>,
]
`;

exports[`Multichannel carousel COMPACT mode dynamic carousel with url buttons 1`] = `
Array [
  <message
    newkey={0}
    type="text"
  >
    *Snatch* - _Five minutes, Turkish_
    
 - Visit website: https://www.imdb.com/title/tt0208092
  </message>,
  <message
    newkey={1}
    type="text"
  >
    *El laberinto del fauno* - _2006_
    
 - Visit website: https://www.imdb.com/title/tt0457430
  </message>,
]
`;
